calico_node/startup.go

```diff
@@ -51,8 +59,239 @@ func main() {
+		valid_ip := net.ParseIP(ip6)
+		if valid_ip.To16() == nil {
+			ip6 = nil
+		}
+	}
+
+	as_num := os.Getenv("AS")
+	if as_num != nil {
+		as, err := numorstring.ASNumberFromString(as_num)
+		if err != nil {
+			as_num = nil
+		}
+	}
+
+	// If ip, ip6 and as_num haven't been set yet, get them from bgp
+	ns, err := c.Nodes().Get(meta)
+	if err != nil {
+		if err == errors.ErrorResourceDoesNotExist {
+			// No other machine has registered configuration under this hostname.
+			// This must be a new host with a unique hostname, which is the
+			// expected behavior.
+		} else {
+			info.Warnf("Error connecting to node: %s", err)
+		}
+
+	} else {
+		bgp_ip := ns.BGP.IPv4Address.String()
+		if ip == nil && bgp_ip != nil {
+			ip = bgp_ip
+		}
+		bgp_ip6 := ns.BGP.IPv6Address.String()
+		if ip6 == nil && bgp_ip6 != nil {
+			ip6 = bgp_ip6
+		}
+		bgp_as_num := ns.BGP.ASNumber.String()
+		if as_num == nil && bgp_as_num != nil {
+			as_num = bgp_as_num
+		}
+		if bgp_ip != nil && bgp_ip != ip {
+			log.Info("WARNING: Hostname '%s' is already in use with IP address %s. 
+				      Calico requires each compute host to have a unique hostname.
+	                  If this is your first time running the Calico node on this host, 
+	                  ensure that another host is not already using the same hostname.", 
+	                  hostname, bgp_ip)
+		}
+	}
+
+	if ip == nil {
+		log.Warnf("Couldn't autodetect a management IPv4 address. Please
+                   provide an IP address either by configuring one in the
+                   node resource, or by re-running the container with the
+                   IP environment variable set.")
+		os.exit(1)
+	} 
+
+    // Write a startup environment file containing the IP address that may have
+    // just been detected.
+    // This is required because the confd templates expect to be able to fill in
+    // some templates by fetching them from the environment.
+	f, err := os.OpenFile("startup.env", os.O_APPEND|os.O_WRONLY, 0600)
+	if err != nil {
+		panic(fmt.Sprintf("Error loading config: %s", err))
+	}
+
+	f.WriteString("export IP=%s\n", ip)
+	f.WriteString("export HOSTNAME=%s\n", hostname)
+
+	warnIfUnknownIp(ip, ip6)
+
+	if strings.ToLower(os.Getenv("NO_DEFAULT_POOLS")) != "true" {
+		poolList, err := c.IPPools().List(api.IPPoolMetadata{})
+		if err != nil {
+			panic(info.Warnf("Unable to fetch IP pool list: %s", err))
+		}
+		ipv4_present, ipv6_present := false, false
+		for p := range poolList {
+			s := strings.Split(p.IPPoolMetadata.CIDR.String())
+			valid_ip := net.ParseIP(ip)
+			if valid_ip.to4() != nil {
+				ipv4_present = true
+			} else {
+				ipv6_present = true
+			}
+			if ipv4_present == true && ipv6_present_present == true {
+				break
+			}
+		}
+
+		if ipv4_present == false {
+			_, cidr, err := net.ParseCIDR("192.168.0.0/16")
+			if err != nil {
+				panic(info.Warnf("Unable to create ipv4 cidr: %s", err))
+			}
+  			ipv4_meta = api.IPPoolMetadata{CIDR: cnet.IPNet{*cidr}}			
```

Comment:
You're missing the Spec section of the IPPool here.  https://godoc.org/github.com/projectcalico/libcalico-go/lib/api#IPPool

We don't do this in the old code, but I think we should be setting the Spec.NATOutgoing field to be true.